<a target="_blank" href="https://colab.research.google.com/github/ZHAW-ZAV/TSO-FS26-students/blob/main/05_linear_regression/05_10_solution_lszh.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

In [5]:
import pandas as pd
import plotly.graph_objects as go
import statsmodels.api as sm

import sys
import os

IN_COLAB = "google.colab" in sys.modules

file_id_lszh = "1JVZm45u6VjoHW26V4ng1k8PEzPyKdXyz"

if IN_COLAB:
    path_to_lszh = "/content/data/lszh_data.csv"
    os.makedirs(os.path.dirname(path_to_lszh), exist_ok=True)
    !gdown "https://drive.google.com/uc?id={file_id_lszh}" -O "{path_to_lszh}"
else:
    import gdown

    url = f"https://drive.google.com/uc?id={file_id_lszh}"
    path_to_lszh = "data/lszh_data.csv"
    os.makedirs(os.path.dirname(path_to_lszh), exist_ok=True)
    gdown.download(url, path_to_lszh, quiet=False)

Downloading...
From: https://drive.google.com/uc?id=1JVZm45u6VjoHW26V4ng1k8PEzPyKdXyz
To: /Users/wate/Documents/GitHub/TSO-FS26-students/05_linear_regression/data/lszh_data.csv
100%|██████████| 1.29k/1.29k [00:00<00:00, 5.49MB/s]


The code above loads the data, do not modify.

***

# TSO Semester Week 5: Linear Regression
In this exercise, we will focus on linear regression models using Python's `statsmodels` library. We will go through the steps of loading data, preprocessing, training a linear regression model, and evaluating its performance. The first part of the exercise is to build a simple linear regression model, while the second part will extend this to a multiple linear regression model.


## Part 1: Simple Linear Regression Model
In the following, we will build a simple linear regression model to predict a target variable $y$ based on a single feature $X$. In order to do so, we will conduct the following steps:
1. Load the data into Python
2. Visualize the data to get a better understanding for the relationship between $X$ and $y$
3. Define the input and output variables of the linear regression model
4. Train the linear regression model
5. Inspect the model
6. Make predictions using the trained model
7. Visualize observations and predictions


### Part 1, Step 1 — Load the Data into Python

**Tasks:**
1. Load the dataset from the provided CSV file into a Pandas DataFrame.
2. Filter the dataset to include only rows in which feature `Year` is less than 2020 (this way, we exclude the COVID-19 pandemic years).

In [9]:
df_full = pd.read_csv(path_to_lszh, delimiter=';')

# Filter the DataFrame for years before 2020
df = df_full.query("Year < 2020")
df

,Year,Population,GDP_per_capita_USD,PAX_per_Year
0,1982,6390000,18015,8367699
1,1983,6427000,17874,8661582
2,1984,6455896,17006,9102355
3,1985,6484834,17167,9546141
4,1986,6523413,24480,9670366
5,1987,6566799,30466,10534015
6,1988,6619973,32690,11305215
7,1989,6673850,31335,12231671
8,1990,6750693,39608,12769757
9,1991,6800000,39578,12225542


### Part 1, Step 2 — Visualize the Relationship between X and y

**Tasks:**
1. Create three different scatter plots to visualize the relationship between the features $X$ (`Year`, `Population`, `GDP_per_capita_USD`) and the target variable $y$ (`PAX_per_Year`).

In [3]:
fig = go.Figure()

fig.add_trace(
    go.Scatter(
        x=df["Year"],
        y=df["PAX_per_Year"],
        mode="markers",
        marker=dict(opacity=0.7),
        name="PAX"
    )
)

fig.update_layout(
    title="PAX per Year vs. Year",
    xaxis_title="Year",
    yaxis_title="PAX per Year",
    template="plotly_white"
)

fig.show()

In [4]:
fig = go.Figure()

fig.add_trace(
    go.Scatter(
        x=df["GDP_per_capita_USD"],
        y=df["PAX_per_Year"],
        mode="markers",
        marker=dict(opacity=0.7),
        name="PAX"
    )
)

fig.update_layout(
    title="PAX per Year vs. GDP per capita USD",
    xaxis_title="GDP per capita USD",
    yaxis_title="PAX per Year",
    template="plotly_white"
)

fig.show()

In [5]:
fig = go.Figure()

fig.add_trace(
    go.Scatter(
        x=df["Population"],
        y=df["PAX_per_Year"],
        mode="markers",
        marker=dict(opacity=0.7),
        name="PAX"
    )
)

fig.update_layout(
    title="PAX per Year vs. Population",
    xaxis_title="Population",
    yaxis_title="PAX per Year",
    template="plotly_white"
)

fig.show()

### Part 1, Step 3 — Define Input (X) and Output (y)

**Tasks:**
1. Define the input variable `X` and the output variable `y` for the linear regression model. Only select one variable / feature for `X` (e.g., `Year`)


In [14]:
feature = "Year"

X = df[feature]
y = df["PAX_per_Year"]

### Part 1, Step 4 — Train a Linear Regression Model

**Tasks:**
1. Use the `statsmodels` library to train a simple linear regression model using the defined input and output variables. Make sure to add the intercept term to $X$ before training the model.

In [ ]:
# statsmodels does NOT add an intercept automatically
X = sm.add_constant(X)

model = sm.OLS(y, X)
results = model.fit()

### Part 1, Step 5 — Inspect the Model with `results.summary()` (statsmodels)

**Tasks:**
1. Use the `summary()` method of the fitted model to inspect the model's.
2. Evaluate the model-level performance metrics (i.e., R-squared, Adjusted R-squared, and F-statistic). What do these metrics indicate about the model's performance?
3. Evaluate the coefficient-level statistics (i.e., coefficients, standard errors, t-values, and p-values). What do these statistics indicate about the significance of the predictor variable?

In [21]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:           PAX_per_Year   R-squared:                       0.934
Model:                            OLS   Adj. R-squared:                  0.932
Method:                 Least Squares   F-statistic:                     507.9
Date:                Wed, 07 Jan 2026   Prob (F-statistic):           8.11e-23
Time:                        10:51:50   Log-Likelihood:                -598.32
No. Observations:                  38   AIC:                             1201.
Df Residuals:                      36   BIC:                             1204.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -1.124e+09   5.07e+07    -22.172      0.0

> ### 📊 Regression Model Summary
>
> **Model-level performance metrics**
>
> - **R-squared (0.934):**  
>   93.4 % of the variance in annual passenger numbers (`PAX_per_Year`) is explained by `Year`, indicating a **very strong linear relationship** between time and passenger demand.
>
> - **Adjusted R-squared (0.932):**  
>   Only marginally lower than the R-squared, as expected for a model with **one explanatory variable**, suggesting no overfitting.
>
> - **F-statistic (507.9):**  
>   The extremely small p-value allows us to **strongly reject the null hypothesis** that the slope coefficient equals zero, indicating the model is **highly statistically significant overall**.
>
> **Coefficient-level statistics**
>
> - **Intercept:**  
>   Represents the predicted passenger volume at `Year = 0`, which has **no meaningful real-world interpretation** in this context and is therefore not substantively relevant.
>
> - **Year:**  
>   Annual passenger numbers increase by approximately **571,000 passengers per year** on average. The large t-value, extremely small p-value, and narrow confidence interval provide **strong evidence that `Year` is a highly statistically significant predictor** of `PAX_per_Year`.

### Part 1, Step 6 — Make Predictions

**Tasks:**
1. Use the trained linear regression model to make predictions on the input data `X`. 
2. Add the predicted values to the original DataFrame as a new column named `predicted_PAX_per_Year`.

In [8]:
df["predicted_PAX_per_Year"] = results.predict(X)
df.head(5)

,Year,Population,GDP_per_capita_USD,PAX_per_Year,predicted_PAX_per_Year
0,1982,6390000,18015,8367699,7.914303e+06
1,1983,6427000,17874,8661582,8.485153e+06
2,1984,6455896,17006,9102355,9.056004e+06
3,1985,6484834,17167,9546141,9.626854e+06
4,1986,6523413,24480,9670366,1.019770e+07


### Part 1, Step 7 — Visualize Model vs Data

**Tasks:**
1. Create a scatter plot of the original data points (actual values) and overlay the regression line (predicted values) to visualize the fit of the model. Plot `Year` on the x-axis and both `PAX_per_Year` and `predicted_PAX_per_Year` on the y-axis.

In [9]:
fig = go.Figure()

# Observed data points
fig.add_trace(
    go.Scatter(
        x=df[feature],
        y=df["PAX_per_Year"],
        mode="markers",
        name="Observed",
        marker=dict(
            symbol="diamond",
            color="red",
            size=8,
            opacity=0.8
        )
    )
)

# Regression line
fig.add_trace(
    go.Scatter(
        x=df[feature],
        y=df["predicted_PAX_per_Year"],
        mode="lines",
        name="Linear Regression Model",
        line=dict(
            color="blue",
            width=2
        )
    )
)

fig.update_layout(
    title="",
    xaxis_title=feature,
    yaxis_title="Annual Passengers Zurich Airport",
    template="plotly_white",
    legend=dict(
        font=dict(size=12)
    )
)

fig.update_xaxes(title_font=dict(size=14))
fig.update_yaxes(title_font=dict(size=14))

fig.show()

# Multiple Linear Regression Model

In [3]:
features = ["Year", "GDP_per_capita_USD", "Population"]

X = df[features]
y = df["PAX_per_Year"]

In [4]:
# statsmodels does NOT add an intercept automatically
X = sm.add_constant(X)

model = sm.OLS(y, X)
results = model.fit()

In [5]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:           PAX_per_Year   R-squared:                       0.950
Model:                            OLS   Adj. R-squared:                  0.945
Method:                 Least Squares   F-statistic:                     214.6
Date:                Wed, 07 Jan 2026   Prob (F-statistic):           3.74e-22
Time:                        11:13:24   Log-Likelihood:                -593.05
No. Observations:                  38   AIC:                             1194.
Df Residuals:                      34   BIC:                             1201.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const              -8.056e+08   2.92

In [6]:
df["predicted_PAX_per_Year"] = results.predict(X)

/var/folders/qb/sbb3fw8j7cx9jqvndn9vkgqw0000gp/T/ipykernel_24721/834851410.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["predicted_PAX_per_Year"] = results.predict(X)


In [7]:
df_pred = df_full.query("Year >= 2020").copy() 
df_pred

,Year,Population,GDP_per_capita_USD,PAX_per_Year
38,2020,8638167,86136,8341047
39,2021,8704546,93967,10234428
40,2022,8777088,94806,22561132
41,2023,8888093,101510,28875199
42,2024,8921981,104681,31195226


In [8]:
X_pred = df_pred[features]
X_pred = sm.add_constant(X_pred)
df_pred["predicted_PAX_per_Year"] = results.predict(X_pred)
df_pred

,Year,Population,GDP_per_capita_USD,PAX_per_Year,predicted_PAX_per_Year
38,2020,8638167,86136,8341047,3.127865e+07
39,2021,8704546,93967,10234428,3.135222e+07
40,2022,8777088,94806,22561132,3.212316e+07
41,2023,8888093,101510,28875199,3.258502e+07
42,2024,8921981,104681,31195226,3.289196e+07


In [16]:
fig = go.Figure()

# Observed data points (pre-2020)
fig.add_trace(
    go.Scatter(
        x=df["Year"],
        y=df["PAX_per_Year"],
        mode="markers",
        name="Observed (pre-2020)",
        marker=dict(
            symbol="diamond",
            color="red",
            size=8,
            opacity=0.8
        )
    )
)

# Observed data points (2020+)
fig.add_trace(
    go.Scatter(
        x=df_pred["Year"],
        y=df_pred["PAX_per_Year"],
        mode="markers",
        name="Observed (2020+)",
        marker=dict(
            symbol="diamond",
            color="red",
            size=8,
            opacity=0.8
        )
    )
)

# Observed data points (2020+)
fig.add_trace(
    go.Scatter(
        x=df_pred["Year"],
        y=df_pred["predicted_PAX_per_Year"],
        mode="lines",
        name="Predicted (2020+)",
        line=dict(
            color="green",
            width=2
        )
    )
)

# Regression line
fig.add_trace(
    go.Scatter(
        x=df["Year"],
        y=df["predicted_PAX_per_Year"],
        mode="lines",
        name="Linear Regression Model",
        line=dict(
            color="blue",
            width=2
        )
    )
)

fig.update_layout(
    title="",
    xaxis_title="Year",
    yaxis_title="Annual Passengers Zurich Airport",
    template="plotly_white",
    legend=dict(
        font=dict(size=12)
    )
)

fig.update_xaxes(title_font=dict(size=14))
fig.update_yaxes(title_font=dict(size=14))

fig.show()